subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c

```diff
@@ -0,0 +1,455 @@
+	/* RPMsg */
+	struct ipc_rpmsg_instance rpmsg_inst;
+
+	/* Static VRINGs */
+	struct ipc_static_vrings vr;
+
+	/* General */
+	struct k_work mbox_work;
+	unsigned int role;
+	atomic_t state;
+};
+
+struct backend_config_t {
+	unsigned int role;
+	uintptr_t shm_addr;
+	size_t shm_size;
+	struct mbox_channel mbox_tx;
+	struct mbox_channel mbox_rx;
+};
+
+static void rpmsg_service_unbind(struct rpmsg_endpoint *ep)
+{
+	rpmsg_destroy_ept(ep);
+}
+
+static struct ipc_rpmsg_ept *get_ept_slot_with_name(struct ipc_rpmsg_instance *rpmsg_inst,
+						    const char *name)
+{
+	struct ipc_rpmsg_ept *rpmsg_ept;
+
+	for (size_t i = 0; i < NUM_ENDPOINTS; i++) {
+		rpmsg_ept = &rpmsg_inst->endpoint[i];
+
+		if (strcmp(name, rpmsg_ept->name) == 0) {
+			return &rpmsg_inst->endpoint[i];
+		}
+	}
+
+	return NULL;
+}
+
+static struct ipc_rpmsg_ept *get_available_ept_slot(struct ipc_rpmsg_instance *rpmsg_inst)
+{
+	return get_ept_slot_with_name(rpmsg_inst, "");
+}
+
+/*
+ * Returns:
+ *  - true:  when the endpoint was already cached / registered
+ *  - false: when the endpoint was never registered before
+ *
+ * Returns in **rpmsg_ept:
+ *  - The endpoint with the name *name if it exists
+ *  - The first endpoint slot available when the endpoint with name *name does
+ *    not exist
+ *  - NULL in case of error
+ */
+static bool get_ept(struct ipc_rpmsg_instance *rpmsg_inst,
+		    struct ipc_rpmsg_ept **rpmsg_ept, const char *name)
+{
+	struct ipc_rpmsg_ept *ept;
+
+	ept = get_ept_slot_with_name(rpmsg_inst, name);
+	if (ept != NULL) {
+		(*rpmsg_ept) = ept;
+		return true;
+	}
+
+	ept = get_available_ept_slot(rpmsg_inst);
+	if (ept != NULL) {
+		(*rpmsg_ept) = ept;
+		return false;
+	}
+
+	(*rpmsg_ept) = NULL;
+
+	return false;
+}
+
+static void advertise_ept(struct ipc_rpmsg_instance *rpmsg_inst, struct ipc_rpmsg_ept *rpmsg_ept,
+			  const char *name, uint32_t dest)
+{
+	struct rpmsg_device *rdev;
+	int err;
+
+	rdev = rpmsg_virtio_get_rpmsg_device(&rpmsg_inst->rvdev);
+
+	err = rpmsg_create_ept(&rpmsg_ept->ep, rdev, name, RPMSG_ADDR_ANY,
+			       dest, rpmsg_inst->cb, rpmsg_service_unbind);
+	if (err != 0) {
+		return;
+	}
+
+	rpmsg_ept->bound = true;
+	if (rpmsg_inst->bound_cb) {
+		rpmsg_inst->bound_cb(rpmsg_ept);
+	}
+}
+
+static void ns_bind_cb(struct rpmsg_device *rdev, const char *name, uint32_t dest)
+{
+	struct ipc_rpmsg_instance *rpmsg_inst;
+	struct rpmsg_virtio_device *p_rvdev;
+	struct ipc_rpmsg_ept *rpmsg_ept;
+	bool ept_cached;
+
+	p_rvdev = CONTAINER_OF(rdev, struct rpmsg_virtio_device, rdev);
+	rpmsg_inst = CONTAINER_OF(p_rvdev->shpool, struct ipc_rpmsg_instance, shm_pool);
+
+	k_mutex_lock(&rpmsg_inst->mtx, K_FOREVER);
+	ept_cached = get_ept(rpmsg_inst, &rpmsg_ept, name);
+
+	if (rpmsg_ept == NULL) {
+		k_mutex_unlock(&rpmsg_inst->mtx);
+		return;
+	}
+
+	if (ept_cached) {
+		/*
+		 * The endpoint was already registered by the HOST core. The
+		 * endpoint can now be advertised to the REMOTE core.
+		 */
+		k_mutex_unlock(&rpmsg_inst->mtx);
```

Comment:
no, it's because it's not useful to keep that locked (AFAICT). The mutex is basically protecting what `get_ept()` is returning (the endpoint), from concurrent access between `ns_bind_cb()` and `register_ept()`.

I don't think that having the mutex locked for `advertise_ept()` is useful (but I could be wrong).

Reply:
no race condition on `rpmsg_ept->bound` between `advertise_ept` and `ept_cb` ?

Reply:
Uhm, I don't think so. `->bound` in `advertise_ept` is exclusively called by the host (either from `ns_bind_cb` or from `register_ept`) while `->bound` in `ept_cb` is only called on the remote. I'll add a comment trying to explain better how this works.